bindings/java/c/host.c

```diff
@@ -0,0 +1,691 @@
+            detach(jenv);
+        }
+
+        //set java method params
+
+        if (!context) {
+            printf("JNI Error: context is NULL in method %s\n",java_method_name);
+            detach(jenv);
+        }
+        jcontext_index = context->index;
+
+        jaddress = (*jenv)->NewByteArray(jenv, ADDRESS_SIZE);
+        (*jenv)->SetByteArrayRegion(jenv, jaddress, 0, ADDRESS_SIZE, (jbyte *)address);
+        
+        if(!jaddress){
+            printf("JNI Error: method params could not be initialized for method: %s\n",java_method_name);
+            detach(jenv);
+        }
+
+        //call java method
+        jint jresult = (*jenv)->CallStaticIntMethod(jenv, host_class, method, jcontext_index, jaddress);
+        result = !!jresult;
+    }
+    return result;
+}
+
+evmc_bytes32 get_storage_fn(struct evmc_host_context* context, const evmc_address* address, const evmc_bytes32* key){
+    evmc_bytes32 result;
+    const char java_method_name[] = "get_storage";
+    const char java_method_signature[] = "(I[B[B)Ljava/nio/ByteBuffer;";
+    JNIEnv *jenv = attach();
+    if(jenv != NULL) {
+        jclass host_class;
+        jmethodID method;
+        jint jcontext_index;
+        jbyteArray jaddress;
+        jbyteArray jkey;
+        //get java class
+        host_class = (*jenv)->FindClass(jenv, "org/ethereum/evmc/Host");
+        if(!host_class){
+            printf("JNI Error: FindClass was unable to load the Host class\n");
+            detach(jenv);
+        }
+    
+        //get java method
+        method = (*jenv)->GetStaticMethodID(jenv, host_class, java_method_name, java_method_signature);
+        if(!method){
+            printf("JNI Error: GetStaticMethodID was unable to find method %s\n",java_method_name);
+            detach(jenv);
+        }
+
+        //set java method params
+
+        if (!context) {
+            printf("JNI Error: context is NULL in method %s\n",java_method_name);
+            detach(jenv);
+        }
+        jcontext_index = context->index;
+
+        jaddress = (*jenv)->NewByteArray(jenv, ADDRESS_SIZE);
+        (*jenv)->SetByteArrayRegion(jenv, jaddress, 0, ADDRESS_SIZE, (jbyte *)address);
+
+        jkey = (*jenv)->NewByteArray(jenv, BYTES32_SIZE);
+        (*jenv)->SetByteArrayRegion(jenv, jkey, 0, BYTES32_SIZE, (jbyte *)key);
+        
+        if(!jaddress || !jkey){
+            printf("JNI Error: method params could not be initialized for method: %s\n",java_method_name);
+            detach(jenv);
+        }
+
+        //call java method
+        jobject jresult = (*jenv)->CallStaticObjectMethod(jenv, host_class, method, jcontext_index, jaddress, jkey);
+
+        if (!jresult) {
+            printf("JNI Error: invalid jresult in method %s\n",java_method_name);
+            detach(jenv);
+        }
+
+        // TODO: need some error handling here
+        result = *(struct evmc_bytes32*) (*jenv)->GetDirectBufferAddress(jenv, jresult);
```

Comment:
I think it would be nicer for `GetDirectBufferAddress` to set a local pointer and than have a separate statement for copying. Then also an assert could be added on the assignment.

e.g.
```
evmc_bytes32 *result_ptr = (struct evmc_bytes32*) (*jenv)->GetDirectBufferAddress(jenv, jresult);
assert(result_ptr != NULL);
result = *result_ptr;
```